### PR TITLE
[file_system_http_cache] Rename stat with tags

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: [file_system_http_cache] Rename stat with tags
Additional Description: Discovered in testing that a stat named `a.b.c` is not the same as a stat that prints out as `a.b.c`, if `b.c` is derived from tags. This changes the stats for `file_system_http_cache` from having the cache path in the stat-name to having the cache path as a tag value.
Risk Level: Very low, extension is WIP.
Testing: Coverage is there; stat-names could be tested better, but right now the fake metrics don't support it correctly. It would make more sense to test stat name generation while creating and migrating to a stats-common method to make tagged stats anyway.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by ravenblackx** [comment]
Stacks on #25386 so that I could write a useful test; for review purposes `isolated_store_*` files can be ignored.

File: source/common/stats/isolated_store_impl.cc
Comment (jmarantz) [inline source/common/stats/isolated_store_impl.cc:43]: Just to make this less verbose could you factor the conditional expression into a static local helper method with signature
`StatNameTagOrVector tagVectorFromOpt(StatNameTagVectorOptConstRef tags);`

File: source/extensions/http/cache/file_system_http_cache/stats.h
Comment (jmarantz) [inline source/extensions/http/cache/file_system_http_cache/stats.h:45]: rvalue-ref not necessary for StatName -- it's just a single pointer. Suggest just passing by value.

Think of StatName like a string_view -- its storage needs to be owned somewhere else, but it's cheap to pass them around.

File: test/common/stats/isolated_store_impl_test.cc
Comment (jmarantz) [inline test/common/stats/isolated_store_impl_test.cc:194]: can you make these tests all include two stats with the same tag-extracted-name but different tags, and make sure they all can be represented?

**Review by jmarantz — COMMENTED** [review]
Many apologies for how long this has taken to review. Just a few minor issues (assuming that the new tests I suggest work)

/wait

**Comment by jmarantz** [comment]
This is great btw -- these are actually not just for tests; these are used in production also, so fixing the tag behavior is important.

File: test/common/stats/isolated_store_impl_test.cc
Comment (ravenblackx) [reply test/common/stats/isolated_store_impl_test.cc:194]: Done (also in the predecessor change that doesn't touch cache, so if you want to merge them separately we can do that).

File: source/extensions/http/cache/file_system_http_cache/stats.h
Comment (ravenblackx) [reply source/extensions/http/cache/file_system_http_cache/stats.h]: Done. Thanks for the analogy, that makes it clearer.

File: source/common/stats/isolated_store_impl.cc
Comment (ravenblackx) [reply source/common/stats/isolated_store_impl.cc]: Done (also in the predecessor PR).

**Review by jmarantz — APPROVED** [review]
Thanks!

As this changes core we should get a senior maintainer sign-off. I think this is the superset PR and it's not too big to review, so let's drop the other one.

/assign-from @envoyproxy/senior-maintainers

**Pull request merged**
